internal/api/middleware/timer.go

```diff
@@ -0,0 +1,46 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"regexp"
+	"strings"
+
+	"github.com/go-chi/chi"
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/dogstatsd"
+)
+
+func getHist(r *http.Request, t *dogstatsd.Timing) metrics.Histogram {
+	str := strings.ReplaceAll(chi.RouteContext(r.Context()).RoutePattern(), "/", ".")
+	fmt.Println(str)
+	regx := regexp.MustCompile(`[{}]`)
+	if regx.MatchString(str) {
+		str = regx.ReplaceAllString(str, "")
+	}
+
+	str = strings.Replace(str, ".", "", 1)
+
+	fmt.Println(str)
+	return t.With(str, "method", r.Method)
+}
+
+func Timer(t *dogstatsd.Timing) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				next.ServeHTTP(w, r)
+				h := getHist(r, t)
+				defer func(t0 time.Time) {
+					h.Observe(millisecondsSince(t0))
+				}(time.Now())
```

Comment:
We need to get the initial time value _before_ calling the next handler.

Reply:
This one is still a thing. The cute `defer` trick likely won't work anymore, since we need a value created before calling the next handler (the initial time) and a value created after calling the next handler (the request context).

Reply:
If I get an initial time before calling the next handler does making use of `defer` after the handler call still work?

```
t0 := time.Now()
next.ServeHTTP(w, r)
id := getIdentifier(r.Context())
defer func(t0 time.Time) {
	h.With("endpoint", id, "method", r.Method).Observe(millisecondsSince(t0))
}(t0)
```